python/private/toolchains_repo.bzl

```diff
@@ -0,0 +1,83 @@
+# Copyright 2022 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Create a repository to hold the toolchains.
+
+This follows guidance here:
+https://docs.bazel.build/versions/main/skylark/deploying.html#registering-toolchains
+
+The "complex computation" in our case is simply downloading large artifacts.
+This guidance tells us how to avoid that: we put the toolchain targets in the
+alias repository with only the toolchain attribute pointing into the
+platform-specific repositories.
+"""
+
+PLATFORMS = {
+    "aarch64-apple-darwin": struct(
+        compatible_with = [
+            "@platforms//os:macos",
+            "@platforms//cpu:aarch64",
+        ],
+    ),
+    "x86_64-apple-darwin": struct(
+        compatible_with = [
+            "@platforms//os:macos",
+            "@platforms//cpu:x86_64",
+        ],
+    ),
+    "x86_64-unknown-linux-gnu": struct(
+        compatible_with = [
+            "@platforms//os:linux",
+            "@platforms//cpu:x86_64",
+        ],
+    ),
+}
+
+def _toolchains_repo_impl(repository_ctx):
+    build_content = """\
+# Generated by toolchains_repo.bzl
+#
+# These can be registered in the workspace file or passed to --extra_toolchains
+# flag. By default all these toolchains are registered by the
+# python_register_toolchains macro so you don't normally need to interact with
+# these targets.
+
+"""
+
+    for [platform, meta] in PLATFORMS.items():
+        build_content += """\
+toolchain(
+    name = "{platform}_toolchain",
+    exec_compatible_with = {compatible_with},
+    target_compatible_with = {compatible_with},
```

Comment:
Is it correct to restrict the toolchain to the same target and exec environment? I would kinda have expected `target` to be left empty since I don't think there is a "target" restriction in the toolchain. Or does the toolchain somehow become embedded in runfiles?

Reply:
I agree this is probably wrong. I think the bug originates in the template
https://github.com/bazel-contrib/rules-template/blob/main/mylang/private/toolchains_repo.bzl#L91-L92

exec_compatible_with toolchain would be used for build actions. If the user wants a python interpreter in the target (like for a docker container) that might be a use case for target_compatible_with, but probably not needed in this PR

Reply:
to be precise, when the target platform differs from the execution platform, I think this code will result in no toolchains being selected

Reply:
Users can totally make their own if they want to constrain a toolchain on some other platforms. I don't think we have to cover long tail use cases, at least until they are requested